homeassistant/components/sonarr/sensor.py

```diff
@@ -86,164 +97,71 @@ async def async_setup_entry(
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up Sonarr sensors based on a config entry."""
-    sonarr: SonarrClient = hass.data[DOMAIN][entry.entry_id][DATA_SONARR]
-    host_config: PyArrHostConfiguration = hass.data[DOMAIN][entry.entry_id][
-        DATA_HOST_CONFIG
+    coordinators: dict[str, SonarrDataUpdateCoordinator] = hass.data[DOMAIN][
+        entry.entry_id
     ]
-    system_status: SystemStatus = hass.data[DOMAIN][entry.entry_id][DATA_SYSTEM_STATUS]
-    options: dict[str, Any] = dict(entry.options)
-
-    entities = [
+    async_add_entities(
         SonarrSensor(
-            sonarr,
-            host_config,
-            system_status,
-            entry.entry_id,
+            coordinators[coordinator_type],
             description,
-            options,
         )
-        for description in SENSOR_TYPES
-    ]
-
-    async_add_entities(entities, True)
-
-
-def sonarr_exception_handler(
-    func: Callable[Concatenate[_SonarrSensorT, _P], Awaitable[None]]
-) -> Callable[Concatenate[_SonarrSensorT, _P], Coroutine[Any, Any, None]]:
-    """Decorate Sonarr calls to handle Sonarr exceptions.
-
-    A decorator that wraps the passed in function, catches Sonarr errors,
-    and handles the availability of the entity.
-    """
-
-    @wraps(func)
-    async def wrapper(
-        self: _SonarrSensorT, *args: _P.args, **kwargs: _P.kwargs
-    ) -> None:
-        try:
-            await func(self, *args, **kwargs)
-            self.last_update_success = True
-        except ArrConnectionException as error:
-            if self.last_update_success:
-                _LOGGER.error("Error communicating with API: %s", error)
-            self.last_update_success = False
-        except ArrException as error:
-            if self.last_update_success:
-                _LOGGER.error("Invalid response from API: %s", error)
-                self.last_update_success = False
-
-    return wrapper
+        for coordinator_type, description in SENSOR_TYPES.items()
```

Comment:
Why change `SENSOR_TYPES` from a tuple to a dict? The `coordinator_type` can just as easily be accessed with `description.key`. With a `dict` you need to be sure to keep the dict key always in sync with the description key.

Reply:
This is same method that is used in Radarr and Lidarr. I am certainly willing tweak it if it makes things better.

Reply:
I would probably choose the tuple myself but the benefit is minor. We can leave the dict if that's more similar to the other integrations.

Reply:
The only issue is what do we do in the future if we decide to have multiple sensors from the same coordinator. We will only have to change it it from `dict[str, description]` to `dict[str, tuple[description]]`.

Reply:
> The only issue is what do we do in the future if we decide to have multiple sensors from the same coordinator. We will only have to change it it from `dict[str, description]` to `dict[str, tuple[description]]`.

Adding a `coordinator_key` attribute might be a bit better but that's nothing we have to worry about now.